# Exploring the Business Dataset

In [1]:
import numpy as np
import pandas as pd
#from time import sleep

In [2]:
%%time
business_df = pd.read_json('../data/business.json', lines=True)
business_df.head()

CPU times: user 2.9 s, sys: 662 ms, total: 3.56 s
Wall time: 3.57 s


business_id                     name              address  \
0  6iYb2HFDywm3zjuRg0shjw      Oskar Blues Taproom         921 Pearl St   
1  tCbdrRPZA0oiIYSmHG3J0w  Flying Elephants at PDX  7000 NE Airport Way   
2  bvN78flM8NLprQ1a1y5dRg           The Reclaimory   4720 Hawthorne Ave   
3  oaepsyvc0J17qwi8cfrOWg              Great Clips   2566 Enterprise Rd   
4  PE9uqAjdw0E4-8mjGl3wVA        Crossfit Terminus  1046 Memorial Dr SE   

          city state postal_code   latitude   longitude  stars  review_count  \
0      Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1     Portland    OR       97218  45.588906 -122.593331    4.0           126   
2     Portland    OR       97214  45.511907 -122.613693    4.5            13   
3  Orange City    FL       32763  28.914482  -81.295979    3.0             8   
4      Atlanta    GA       30316  33.747027  -84.353424    4.0            14   

   is_open                                         attributes  \
0        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
1        1  {'RestaurantsTakeOut': 'True', 'RestaurantsAtt...   
2        1  {'BusinessAcceptsCreditCards': 'True', 'Restau...   
3        1  {'RestaurantsPriceRange2': '1', 'BusinessAccep...   
4        1  {'GoodForKids': 'False', 'BusinessParking': '{...   

                                          categories  \
0  Gastropubs, Food, Beer Gardens, Restaurants, B...   
1  Salad, Soup, Sandwiches, Delis, Restaurants, C...   
2  Antiques, Fashion, Used, Vintage & Consignment...   
3                         Beauty & Spas, Hair Salons   
4  Gyms, Active Life, Interval Training Gyms, Fit...   

                                               hours  
0  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
1  {'Monday': '5:0-18:0', 'Tuesday': '5:0-17:0', ...  
2  {'Thursday': '11:0-18:0', 'Friday': '11:0-18:0...  
3                                               None  
4  {'Monday': '16:0-19:0', 'Tuesday': '16:0-19:0'...

In [3]:
business_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 160585 entries, 0 to 160584
Data columns (total 14 columns):
 #   Column        Non-Null Count   Dtype  
---  ------        --------------   -----  
 0   business_id   160585 non-null  object 
 1   name          160585 non-null  object 
 2   address       160585 non-null  object 
 3   city          160585 non-null  object 
 4   state         160585 non-null  object 
 5   postal_code   160585 non-null  object 
 6   latitude      160585 non-null  float64
 7   longitude     160585 non-null  float64
 8   stars         160585 non-null  float64
 9   review_count  160585 non-null  int64  
 10  is_open       160585 non-null  int64  
 11  attributes    145593 non-null  object 
 12  categories    160470 non-null  object 
 13  hours         133244 non-null  object 
dtypes: float64(3), int64(2), object(9)
memory usage: 17.2+ MB


In [4]:
business_df = business_df.assign(categories = business_df['categories'].str.split(', ')).explode('categories')
business_df

business_id                 name              address  \
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
0       6iYb2HFDywm3zjuRg0shjw  Oskar Blues Taproom         921 Pearl St   
...                        ...                  ...                  ...   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   
160584  ngmLL5Y5OT-bYHKU0kKrYA          Zora Grille  1370 E Altamonte Dr   

                     city state postal_code   latitude   longitude  stars  \
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
0                 Boulder    CO       80302  40.017544 -105.283348    4.0   
...                   ...   ...         ...        ...         ...    ...   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   
160584  Altamonte Springs    FL       32701  28.663047  -81.346728    4.5   

        review_count  is_open  \
0                 86        1   
0                 86        1   
0                 86        1   
0                 86        1   
0                 86        1   
...              ...      ...   
160584            82        0   
160584            82        0   
160584            82        0   
160584            82        0   
160584            82        0   

                                               attributes       categories  \
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...       Gastropubs   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...             Food   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...     Beer Gardens   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...      Restaurants   
0       {'RestaurantsTableService': 'True', 'WiFi': 'u...             Bars   
...                                                   ...              ...   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...      Restaurants   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...   Middle Eastern   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...    Mediterranean   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...  Persian/Iranian   
160584  {'RestaurantsReservations': 'True', 'OutdoorSe...            Salad   

                                                    hours  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
0       {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
...                                                   ...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  
160584  {'Tuesday': '17:0-21:0', 'Wednesday': '17:0-21...  

[709083 rows x 14 columns]

Note that the method above will actually create additional rows which are essentially duplicates, except that each row will contain one of the categories that were previously grouped together. We will deal with these duplicate rows later. Let us double check the shape of our dataset - the number of rows should have increased.

In [5]:
business_df.shape

(709083, 14)

As expected, the number of rows has increased from 160,585 to 709,083.

Let us examine the number of different categories in this dataset.

In [10]:
len(business_df['categories'].unique())

1331

There are 1,331 unique business categories in this dataset! Let's look at the top 40 categories.

In [11]:
business_df['categories'].value_counts().head(40)

Restaurants                  50763
Food                         29469
Shopping                     26205
Beauty & Spas                16574
Home Services                16465
Health & Medical             15102
Local Services               12192
Nightlife                    11990
Bars                         10741
Automotive                   10119
Event Planning & Services     9644
Active Life                   9231
Coffee & Tea                  7725
Sandwiches                    7272
Fashion                       6599
American (Traditional)        6541
Hair Salons                   5900
Pizza                         5756
Hotels & Travel               5703
Breakfast & Brunch            5505
Home & Garden                 5499
American (New)                5458
Real Estate                   5361
Arts & Entertainment          5314
Fast Food                     5300
Auto Repair                   5039
Professional Services         4845
Fitness & Instruction         4741
Doctors             

We now filter for establishments with the category names of restaurant, food, brunch, and breakfast as save those establishments to a new dataframe.

In [12]:
restaurant_df = (business_df[business_df['categories'].str.contains('restaurant|food|brunch|breakfast', case=False) == True]).reset_index(drop=True)
restaurant_df.head()

business_id                     name              address  \
0  6iYb2HFDywm3zjuRg0shjw      Oskar Blues Taproom         921 Pearl St   
1  6iYb2HFDywm3zjuRg0shjw      Oskar Blues Taproom         921 Pearl St   
2  tCbdrRPZA0oiIYSmHG3J0w  Flying Elephants at PDX  7000 NE Airport Way   
3  D4JtQNTI4X3KcbzacDJsMw      Bob Likes Thai Food         3755 Main St   
4  jFYIsSb7r1QeESVUnXPHBw          Boxwood Biscuit        740 S High St   

        city state postal_code   latitude   longitude  stars  review_count  \
0    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
2   Portland    OR       97218  45.588906 -122.593331    4.0           126   
3  Vancouver    BC         V5V  49.251342 -123.101333    3.5           169   
4   Columbus    OH       43206  39.947007  -82.997471    4.5            11   

   is_open                                         attributes  \
0        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
1        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
2        1  {'RestaurantsTakeOut': 'True', 'RestaurantsAtt...   
3        1  {'GoodForKids': 'True', 'Alcohol': 'u'none'', ...   
4        1                                               None   

           categories                                              hours  
0                Food  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
1         Restaurants  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
2         Restaurants  {'Monday': '5:0-18:0', 'Tuesday': '5:0-17:0', ...  
3         Restaurants  {'Monday': '17:0-21:0', 'Tuesday': '17:0-21:0'...  
4  Breakfast & Brunch     {'Saturday': '8:0-14:0', 'Sunday': '8:0-14:0'}

In [13]:
restaurant_df.shape

(106453, 14)

In [14]:
len(restaurant_df['categories'].unique())

26

There are less unique number of categories in our dataset. Let's see what these are.

In [15]:
restaurant_df['categories'].unique()

array(['Food', 'Restaurants', 'Breakfast & Brunch', 'Food Court',
       'Seafood', 'Food Delivery Services', 'Food Trucks',
       'Specialty Food', 'Ethnic Food', 'Fast Food', 'Food Stands',
       'Do-It-Yourself Food', 'Seafood Markets', 'Bed & Breakfast',
       'Comfort Food', 'Soul Food', 'Food Safety Training', 'Food Tours',
       'Live/Raw Food', 'Restaurant Supplies', 'Imported Food',
       'Pop-Up Restaurants', 'Food Banks', 'Swiss Food',
       'Dance Restaurants', 'Pub Food'], dtype=object)

In [ ]:
restaurant_df['categories'].value_counts()

Restaurants               50763
Food                      29469
Breakfast & Brunch         5505
Fast Food                  5300
Specialty Food             4272
Seafood                    2958
Food Trucks                2537
Food Stands                1321
Food Delivery Services      960
Ethnic Food                 859
Comfort Food                721
Soul Food                   323
Bed & Breakfast             265
Seafood Markets             249
Imported Food               247
Food Court                  214
Do-It-Yourself Food         188
Live/Raw Food               118
Food Tours                   67
Pop-Up Restaurants           62
Restaurant Supplies          33
Food Banks                   16
Food Safety Training          2
Pub Food                      2
Swiss Food                    1
Dance Restaurants             1
Name: categories, dtype: int64

There are some categories here which are not related to our analysis. We will drop any rows associated with these categories below.

In [17]:
categories_to_drop = ['Food Safety Training','Food Trucks', 'Food Delivery Services',
                      'Food Stands', 'Food Court','Restaurant Supplies', 'Food Banks',
                      'Seafood Markets','Imported Food']

# Save the rows that are not categorized by the labels in categories_to_drop
restaurant_df = restaurant_df[~restaurant_df['categories'].isin(categories_to_drop)]
restaurant_df.shape

(100874, 14)

As mentioned earlier, the explode() method created duplicate rows for each business with unique category identifiers. Let's drop those duplicates now.

In [18]:
restaurant_df = restaurant_df.drop_duplicates(subset=['business_id'], ignore_index=True)
restaurant_df.head()

business_id                      name              address  \
0  6iYb2HFDywm3zjuRg0shjw       Oskar Blues Taproom         921 Pearl St   
1  tCbdrRPZA0oiIYSmHG3J0w   Flying Elephants at PDX  7000 NE Airport Way   
2  D4JtQNTI4X3KcbzacDJsMw       Bob Likes Thai Food         3755 Main St   
3  jFYIsSb7r1QeESVUnXPHBw           Boxwood Biscuit        740 S High St   
4  rYs_1pNB_RMtn5WQh55QDA  Chautauqua General Store      100 Clematis Dr   

        city state postal_code   latitude   longitude  stars  review_count  \
0    Boulder    CO       80302  40.017544 -105.283348    4.0            86   
1   Portland    OR       97218  45.588906 -122.593331    4.0           126   
2  Vancouver    BC         V5V  49.251342 -123.101333    3.5           169   
3   Columbus    OH       43206  39.947007  -82.997471    4.5            11   
4    Boulder    CO       80302  39.998449 -105.281006    3.5             5   

   is_open                                         attributes  \
0        1  {'RestaurantsTableService': 'True', 'WiFi': 'u...   
1        1  {'RestaurantsTakeOut': 'True', 'RestaurantsAtt...   
2        1  {'GoodForKids': 'True', 'Alcohol': 'u'none'', ...   
3        1                                               None   
4        1  {'BikeParking': 'True', 'RestaurantsTakeOut': ...   

           categories                                              hours  
0                Food  {'Monday': '11:0-23:0', 'Tuesday': '11:0-23:0'...  
1         Restaurants  {'Monday': '5:0-18:0', 'Tuesday': '5:0-17:0', ...  
2         Restaurants  {'Monday': '17:0-21:0', 'Tuesday': '17:0-21:0'...  
3  Breakfast & Brunch     {'Saturday': '8:0-14:0', 'Sunday': '8:0-14:0'}  
4                Food  {'Monday': '8:0-20:0', 'Tuesday': '8:0-20:0', ...

In [19]:
restaurant_df.shape

(64079, 14)

We now have 64,079 restaurants in this dataframe.

In [20]:
len(restaurant_df['business_id'].unique())

64079

In [34]:
%%time
restaurant_df = pd.concat([restaurant_df, restaurant_df['attributes'].apply(pd.Series)], axis=1).drop('attributes', axis=1)

CPU times: user 15.6 s, sys: 161 ms, total: 15.8 s
Wall time: 15.8 s


In [35]:
restaurant_df.shape

(64079, 52)

In [36]:
restaurant_df.columns

Index(['business_id', 'name', 'address', 'city', 'state', 'postal_code',
       'latitude', 'longitude', 'stars', 'review_count', 'is_open',
       'categories', 'hours', 'RestaurantsTableService', 'WiFi', 'BikeParking',
       'BusinessParking', 'BusinessAcceptsCreditCards',
       'RestaurantsReservations', 'WheelchairAccessible', 'Caters',
       'OutdoorSeating', 'RestaurantsGoodForGroups', 'HappyHour',
       'BusinessAcceptsBitcoin', 'RestaurantsPriceRange2', 'Ambience', 'HasTV',
       'Alcohol', 'GoodForMeal', 'DogsAllowed', 'RestaurantsTakeOut',
       'NoiseLevel', 'RestaurantsAttire', 'RestaurantsDelivery', 'GoodForKids',
       'ByAppointmentOnly', 'Music', 'GoodForDancing', 'BestNights', 'BYOB',
       'CoatCheck', 'Smoking', 'DriveThru', 'BYOBCorkage', 'Corkage',
       'RestaurantsCounterService', 'AcceptsInsurance', 'DietaryRestrictions',
       'AgesAllowed', 'Open24Hours', 'HairSpecializesIn'],
      dtype='object')